sample/android-app/src/main/java/com/icerockdev/MainActivity.kt

```diff
@@ -1,13 +1,74 @@
 package com.icerockdev
 
 import android.os.Bundle
+import android.view.View
+import android.widget.LinearLayout
+import android.widget.Toast
 import androidx.appcompat.app.AppCompatActivity
+import com.google.android.material.snackbar.Snackbar
+import dev.icerock.moko.permissions.DeniedAlwaysException
+import dev.icerock.moko.permissions.DeniedException
+import dev.icerock.moko.permissions.Permission
+import dev.icerock.moko.permissions.PermissionsController
+import android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS
+import android.content.Intent
+import android.net.Uri
+import com.icerockdev.library.SampleViewModel
+import dev.icerock.moko.mvvm.dispatcher.eventsDispatcherOnMain
 
-class MainActivity : AppCompatActivity() {
+class MainActivity : AppCompatActivity(), SampleViewModel.EventListener {
+
+    private lateinit var sampleViewModel: SampleViewModel
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-
         setContentView(R.layout.activity_main)
+
+        // Prepares the permissions controller and binds it to the activity lifecycle.
+        val permissionsController = PermissionsController().also {
+            it.bind(lifecycle, supportFragmentManager)
+        }
+
+        // Creates viewModel from common code.
+        sampleViewModel = SampleViewModel(
+            eventsDispatcher = eventsDispatcherOnMain(),
+            permissionsController = permissionsController
+        ).also {
+            it.eventsDispatcher.bind(this, this)
+        }
+    }
+
+    fun onRequestButtonClick(view: View?) {
+        // Starts permission providing process.
+        sampleViewModel.providePermission(Permission.CAMERA)
```

Comment:
to follow MVVM let's call a method `viewModel.onRequestButtonPressed()`.

Reply:
Renamed.